iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,153 @@
+    for (auto& element : m_serviceDescriptionVector)
+    {
+        if (element.first == serviceDescription)
+        {
+            // Due to n:m communication we don't store twice but increase the reference counter
+            element.second++;
+            return cxx::success<>();
+        }
+    }
+    uint64_t referenceCounter = 1;
+    if (!m_serviceDescriptionVector.push_back({serviceDescription, referenceCounter}))
+    {
+        return cxx::error<Error>(Error::SERVICE_REGISTRY_FULL);
+    }
+
+    auto serviceIndex = m_serviceDescriptionVector.size() - 1;
+    m_serviceMap.insert({serviceDescription.getServiceIDString(), serviceIndex});
+    m_instanceMap.insert({serviceDescription.getInstanceIDString(), serviceIndex});
+    return cxx::success<>();
 }
 
-void ServiceRegistry::remove(const capro::IdString_t& service, const capro::IdString_t& instance)
+bool ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::remove(m_serviceMap[service].instanceSet, instance);
+    bool removedElement{false};
+    uint64_t index = 0U;
+    for (auto iterator = m_serviceDescriptionVector.begin(); iterator != m_serviceDescriptionVector.end();)
+    {
+        auto& element = m_serviceDescriptionVector[index].first;
+        if (element == serviceDescription)
+        {
+            auto& refCounter = m_serviceDescriptionVector[index].second;
+            refCounter--;
+            if (refCounter == 0)
+            {
+                m_serviceDescriptionVector.erase(iterator);
+                removedElement = true;
+            }
+            else
+            {
+                return true;
+            }
+            // There can't be more than one element
+            break;
+        }
+        index++;
+        iterator++;
+    }
+
+    auto removeIndexFromMap = [](std::multimap<capro::IdString_t, uint64_t>& map, uint64_t index) {
+        for (auto it = map.begin(); it != map.end();)
+        {
+            if (it->second == index)
+            {
+                it = map.erase(it);
+                continue;
+            }
+            else if (it->second > index)
+            {
+                // update index due to removed element
+                it->second--;
+            }
+            it++;
+        }
+    };
+
+    removeIndexFromMap(m_serviceMap, index);
+    removeIndexFromMap(m_instanceMap, index);
+
+    return removedElement;
 }
 
-void ServiceRegistry::find(InstanceSet_t& instances,
+void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const capro::IdString_t& service,
-                           const capro::IdString_t& instance) const
+                           const capro::IdString_t& instance) const noexcept
 {
-    if (instance == Wildcard)
+    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
+
+    // Find (K1, K2)
+    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
```

Comment:
You can write it much simpler, it is just: `O(log(n))` 

Here is why:
```
log n + log n = log ( n * n ) = log ( n ^ 2 )
log ( n ^ 2 ) = 2 * log n
```

See: https://en.wikipedia.org/wiki/List_of_logarithmic_identities

And the big-O-notation: https://en.wikipedia.org/wiki/Big_O_notation states
that `O(log(n))` that the `Find(K1, K2)` function has an upper limit which is `log(n)` with some constant.
`Find(K1, K2) = O(log(n)) => Find(K1, K2) <= M * log(n)` with `M > 0`. So the `M` constant contains the factor 2 and all the constants with `max(#PossibleServices + #possibleInstances) + #intersection)`.

Reply:
We hence can simplify this to `O( log n + #resultset)`, but not `O(log n)`.